### DISCUSSION TITLE
Hosting the CLR

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/runtime#99210] [Q&A] [answered]
Is there any kind of specification for what a platform must provide in order to host the CLR and thus .Net applications? Clearly the CLR expects some kind of threading facility and perhaps a file system, but what else does it need? what is the minimum a platform must have for the CLR to work?

If one had a bare machine without even an operating system, what would they need to have in place for the CLR to be hosted?

**Comment by alexrp** [comment]
> Is there any kind of specification for what a platform must provide in order to host the CLR and thus .Net applications?

In the abstract sense, Ecma 335 specifies the facilities needed for a conforming CLI implementation. But this is likely not what you're interested in. Concretely, both CoreCLR and Mono expect (in no particular order): Threading, processes, file system, sockets/networking, virtual memory, console/terminal access, date/time and timer features, environment variables, POSIX signals, etc.

> If one had a bare machine without even an operating system, what would they need to have in place for the CLR to be hosted?

Basically, an operating system. 🙂

**Reply by The-Futurist** [reply]
OK thanks, this is a start!

When you say _virtual memory,_ in what sense? the presence of virtual memory is (ordinarily) completely invisible to application code isn't it?

Also, if there was not any sockets/networking, obviously many network based classes could not work, but is there anything more fundamental that would be impacted? Could one have an OS that did not (in an early version anyway) have sockets yet still run the CLR provided no user code tried to do networking?

**Reply by alexrp** [reply, marked as answer]
> When you say virtual memory, in what sense?

`VirtualAlloc`/`mmap`

> the presence of virtual memory is (ordinarily) completely invisible to application code isn't it?

Yes, but the JIT and GC need it.

> Also, if there was not any sockets/networking, obviously many network based classes could not work, but is there anything more fundamental that would be impacted? Could one have an OS that did not (in an early version anyway) have sockets yet still run the CLR provided no user code tried to do networking?

I don't see any reason why it wouldn't be workable. It might require some build system plumbing and `[UnsupportedOSPlatform(...)]` annotations, though. I *think* (but don't quote me on this) that `browser-wasm` is a bit like this, i.e. I don't think you get access to raw sockets in that target.

**Reply by The-Futurist** [reply]
So does the CLR/runtime (which is the correct term these days?) rely on interop to interact with the host or does it need other special stuff setting up? Imagine we had a bare metal setup on some machine, I'm trying to get an idea what we'd need to write or put in place in order to get the runtime up and running, to achieve the equivalent of this:

```
  dotnet "Demo App\bin\Debug\netcoreapp1.0\Demo App.dll"  
```

that example itself implies some kind of file system, in some form or other.

**Comment by neon-sunset** [comment]
> If one had a bare machine without even an operating system, what would they need to have in place for the CLR to be hosted?

Not sure if this addresses your concerns but take a look at https://github.com/bflattened/bflat which is capable of producing UEFI binaries directly.

**Reply by The-Futurist** [reply]
It might, it is an interesting idea. I'm exploring alternatives to having an OS (like Linux) on STM32MP boards. In theory one could just run pure .Net code (of course that code would be doing all of the hardware control). Today it seems that an OS is needed to host the CLR and that OS also comes with all kinds of support for various hardware devices (like SPI for example). If one could dispense with that it might be possible to create a very elegant and clean ecosystem...

**Comment by The-Futurist** [comment]
@alexrp 
I've been looking at .Net running on [certain STM32 boards](https://www.ghielectronics.com/endpoint/) and although this has been done for Windows and Linux I'm interested in exactly how one could host it without these operating systems. Can anyone explain how the runtime for Linux and Windows, differ? what parts of the runtime are independent of OS and what parts are dependent on it? I'm curious about whether it might be feasible to create a minimal operating system that has enough functionality to support .Net, trying to get a handle on the scope of such an undertaking.

That Endpoint board hosts some build of Linux and I'm no fan of Linux and want to get away from Linux imposing it's own abstractions of the STM32 peripherals which is somewhat contrived and arbitrary, for example we can't use DMA from C# in any direct way because Linux denies access to the DMA resources (but does not deny access to other resources like GPIO or Timers).

I'm not unfamiliar with bare metal STM32 coding so there might be scope for a project that can deliver a core OS that is enough for .Net to run and avoid all of the Linux baggage entirely.

I'm [reading through this ](https://learn.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting)but would love to get any more details or tips about this.

**Reply by huoyaoyuan** [reply]
It should be [.NET IoT](https://dotnet.microsoft.com/apps/iot) as `System.Devices` appears in the video.

**Reply by snikeguo** [reply]
> @alexrp I've been looking at .Net running on [certain STM32 boards](https://www.ghielectronics.com/endpoint/) and although this has been done for Windows and Linux I'm interested in exactly how one could host it without these operating systems. Can anyone explain how the runtime for Linux and Windows, differ? what parts of the runtime are independent of OS and what parts are dependent on it? I'm curious about whether it might be feasible to create a minimal operating system that has enough functionality to support .Net, trying to get a handle on the scope of such an undertaking.
> 
> That Endpoint board hosts some build of Linux and I'm no fan of Linux and want to get away from Linux imposing it's own abstractions of the STM32 peripherals which is somewhat contrived and arbitrary, for example we can't use DMA from C# in any direct way because Linux denies access to the DMA resources (but does not deny access to other resources like GPIO or Timers).
> 
> I'm not unfamiliar with bare metal STM32 coding so there might be scope for a project that can deliver a core OS that is enough for .Net to run and avoid all of the Linux baggage entirely.
> 
> I'm [reading through this ](https://learn.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting)but would love to get any more details or tips about this.

http://icsharptech.com/
.net10 nativeaot for stm32(armv7m)
